### DISCUSSION TITLE
axis error when using predict() on random survival forest

### DISCUSSION THREAD

**Discussion by cchanzl** [sebp/scikit-survival#209] [Q&A]
Hi, I was able to replicate the results in the example (https://scikit-survival.readthedocs.io/en/latest/user_guide/random-survival-forest.html) without issue.

However, when I used my own dataset, I encountered the error "AxisError: axis 1 is out of bounds for array of dimension 1", even though the type of the array used in predict() is the same. There was no issue fitting the rsf.

The dimensions of the array used in predict is also aligned with the training set (ie same number of features).

The full traceback is below.

Does anyone have an idea why this might be the case?

---------------------------------------------------------------------------
AxisError                                 Traceback (most recent call last)
<ipython-input-12-eae85a70579b> in <module>
----> 1 pred_risk = rsf.predict(rsf_x_train)

~\AppData\Roaming\Python\Python37\site-packages\sksurv\ensemble\forest.py in predict(self, X)
    398             Predicted risk scores.
    399         """
--> 400         return self._predict("predict", X)
    401 
    402     def predict_cumulative_hazard_function(self, X, return_array=False):

~\AppData\Roaming\Python\Python37\site-packages\sksurv\ensemble\forest.py in _predict(self, predict_fn, X)
    368                  **_joblib_parallel_args(require="sharedmem"))(
    369             delayed(_accumulate_prediction)(_get_fn(e, predict_fn), X, [y_hat], lock)
--> 370             for e in self.estimators_)
    371 
    372         y_hat /= len(self.estimators_)

~\anaconda3\lib\site-packages\joblib\parallel.py in __call__(self, iterable)
   1015 
   1016             with self._backend.retrieval_context():
-> 1017                 self.retrieve()
   1018             # Make sure that we get a last message telling us we are done
   1019             elapsed_time = time.time() - self._start_time

~\anaconda3\lib\site-packages\joblib\parallel.py in retrieve(self)
    907             try:
    908                 if getattr(self._backend, 'supports_timeout', False):
--> 909                     self._output.extend(job.get(timeout=self.timeout))
    910                 else:
    911                     self._output.extend(job.get())

~\anaconda3\lib\multiprocessing\pool.py in get(self, timeout)
    655             return self._value
    656         else:
--> 657             raise self._value
    658 
    659     def _set(self, i, obj):

~\anaconda3\lib\multiprocessing\pool.py in worker(inqueue, outqueue, initializer, initargs, maxtasks, wrap_exception)
    119         job, i, func, args, kwds = task
    120         try:
--> 121             result = (True, func(*args, **kwds))
    122         except Exception as e:
    123             if wrap_exception and func is not _helper_reraises_exception:

~\anaconda3\lib\site-packages\joblib\_parallel_backends.py in __call__(self, *args, **kwargs)
    606     def __call__(self, *args, **kwargs):
    607         try:
--> 608             return self.func(*args, **kwargs)
    609         except KeyboardInterrupt:
    610             # We capture the KeyboardInterrupt and reraise it as

~\anaconda3\lib\site-packages\joblib\parallel.py in __call__(self)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~\anaconda3\lib\site-packages\joblib\parallel.py in <listcomp>(.0)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~\anaconda3\lib\site-packages\sklearn\ensemble\_forest.py in _accumulate_prediction(predict, X, out, lock)
    465     complains that it cannot pickle it when placed there.
    466     """
--> 467     prediction = predict(X, check_input=False)
    468     with lock:
    469         if len(out) == 1:

~\AppData\Roaming\Python\Python37\site-packages\sksurv\tree\tree.py in predict(self, X, check_input)
    382         """
    383         chf = self.predict_cumulative_hazard_function(X, check_input, return_array=True)
--> 384         return chf.sum(1)
    385 
    386     def predict_cumulative_hazard_function(self, X, check_input=True, return_array=False):

~\AppData\Roaming\Python\Python37\site-packages\numpy\core\_methods.py in _sum(a, axis, dtype, out, keepdims, initial, where)
     45 def _sum(a, axis=None, dtype=None, out=None, keepdims=False,
     46          initial=_NoValue, where=True):
---> 47     return umr_sum(a, axis, dtype, out, keepdims, initial, where)
     48 
     49 def _prod(a, axis=None, dtype=None, out=None, keepdims=False,

AxisError: axis 1 is out of bounds for array of dimension 1

**Comment by sebp** [comment]
It seems that predict_cumulative_hazard_function in SurvivalTree is not 
returning a 2d-array for some reason. The second dimension (axis 1) 
should be time, but is not present for some reason.

One scenario I could imagine is when almost all data are censored.

If run your code in a debugger, could you please check what the value of 
  self.event_times_ inside predict_cumulative_hazard_function of 
SurvivalTree is?

**Reply by cchanzl** [reply]
thanks for the answer, it helped where I was stuck for the past 24 hours.

What you said was correct, the issue was mainly in my y array. Although I had both columns "event" and "time", what I actually had in "time" was "remaining useful life" instead of the complementary "time survived until" required.

Changing that fixed predict() and now score() and predict()  work fine.

thank you so much again and for the great work on this package.

**Reply by thomaslandman** [reply]
I have the same problem and you as you said I indeed have a dataset where almost all data is censored. In my case self.event_times_ is array([0.]).

Do you have any idea how I can fix this?

**Reply by sebp** [reply]
> I have the same problem and you as you said I indeed have a dataset where almost all data is censored. In my case self.event_times_ is array([0.]).
> 
> Do you have any idea how I can fix this?

I'm afraid if your data is so severely censored, there just isn't enough (exact) information about event times to learn from, or evaluate on.
